File: implementations/go/cmd/saga-go/lexer.go
package main

import (
	"fmt"
	"strings"
	"unicode"
	"unicode/utf8"
)

type Lexer struct {
	src            []rune
	pos, line, col int
	file           string
}

func lex(src, file string) ([]Token, error) {
	if !utf8.ValidString(src) {
		return nil, &SagaError{"SAGA-L001", "SAGA-L104", "source is not valid UTF-8", file, 1, 1}
	}
	l := &Lexer{src: []rune(src), line: 1, col: 1, file: file}
	out := []Token{}
	for !l.end() {
		c := l.peek()
		line, col := l.line, l.col
		if unicode.IsSpace(c) {
			l.advance()
			continue
		}
		if c == '#' || (c == '/' && l.peekN(1) == '/') {
			for !l.end() && l.peek() != '\n' {
				l.advance()
			}
			continue
		}
		if isBidiControl(c) {
			return nil, &SagaError{"SAGA-L001", "SAGA-L106", "bidirectional control character is not allowed outside strings", file, line, col}
		}
		if isStart(c) {
			start := l.pos
			l.advance()
			for !l.end() && isContinue(l.peek()) {
				l.advance()
			}
			text := string(l.src[start:l.pos])
			if normalizeNFC(text) != text {
				return nil, &SagaError{"SAGA-L001", "SAGA-L105", "identifier is not NFC-normalized", file, line, col}
			}
			k := IDENT
			if q, ok := keywords[text]; ok {
				k = q
			}
			out = append(out, Token{k, text, line, col, file})
			continue
		}
		if c >= '0' && c <= '9' {
			start := l.pos
			l.advance()
			for !l.end() && ((l.peek() >= '0' && l.peek() <= '9') || l.peek() == '_') {
				l.advance()
			}
			k := INTLIT
			if !l.end() && l.peek() == '.' && l.peekN(1) == '_' {
				return nil, &SagaError{"SAGA-L001", "SAGA-L103", "underscore cannot immediately follow decimal point", file, line, col}
			}
			if !l.end() && l.peek() == '.' && l.peekN(1) != '.' && l.peekN(1) >= '0' && l.peekN(1) <= '9' {
				k = DECLIT
				l.advance()
				for !l.end() && ((l.peek() >= '0' && l.peek() <= '9') || l.peek() == '_') {
					l.advance()
				}
			}
			raw := string(l.src[start:l.pos])
			for _, part := range strings.Split(raw, ".") {
				if part == "" || strings.HasPrefix(part, "_") || strings.HasSuffix(part, "_") || strings.Contains(part, "__") {
					return nil, &SagaError{"SAGA-L001", "SAGA-L103", "numeric separators must occur singly between ASCII digits", file, line, col}
				}
			}
			// Edition 2027 adds explicit IEEE literals. A suffix is required so the
			// existing exact decimal semantics remain unchanged.
			if !l.end() && l.peek() == 'f' {
				if l.peekN(1) == '3' && l.peekN(2) == '2' {
					l.advance()
					l.advance()
					l.advance()
					raw += "f32"
					k = FLOAT32LIT
				} else if l.peekN(1) == '6' && l.peekN(2) == '4' {
					l.advance()
					l.advance()
					l.advance()
					raw += "f64"
					k = FLOAT64LIT
				}
			}
			out = append(out, Token{k, raw, line, col, file})
			continue
		}
		if unicode.IsDigit(c) {
			return nil, &SagaError{"SAGA-L001", "SAGA-L103", "numeric literals use ASCII digits 0-9", file, line, col}
		}
		if c == '$' && (l.peekN(1) == '"' || l.peekN(1) == '\'') {
			l.advance()
			q := l.advance()
			var b strings.Builder
			for !l.end() && l.peek() != q {
				r := l.advance()
				if r == '\\' {
					if l.end() {
						return nil, &SagaError{"SAGA-L001", "SAGA-L102", "unterminated interpolated string escape", file, line, col}
					}
					e := l.advance()
					switch e {
					case 'n':
						b.WriteRune('\n')
					case 't':
						b.WriteRune('\t')
					case 'r':
						b.WriteRune('\r')
					case '\\', '"', '\'':
						b.WriteRune(e)
					case '$':
						b.WriteRune('$')
					default:
						return nil, &SagaError{"SAGA-L001", "SAGA-L101", fmt.Sprintf("unsupported escape \\%c", e), file, line, col}
					}
				} else {
					b.WriteRune(r)
				}
			}
			if l.end() {
				return nil, &SagaError{"SAGA-L001", "SAGA-L102", "unterminated interpolated string", file, line, col}
			}
			l.advance()
			out = append(out, Token{INTERPSTRING, b.String(), line, col, file})
			continue
		}
		if c == '"' || c == '\'' {
			q := c
			l.advance()
			var b strings.Builder
			for !l.end() && l.peek() != q {
				r := l.advance()
				if r == '\\' {
					if l.end() {
						return nil, &SagaError{"SAGA-L001", "SAGA-L102", "unterminated string escape", file, line, col}
					}
					e := l.advance()
					switch e {
					case 'n':
						b.WriteRune('\n')
					case 't':
						b.WriteRune('\t')
					case 'r':
						b.WriteRune('\r')
					case '\\', '"', '\'':
						b.WriteRune(e)
					default:
						return nil, &SagaError{"SAGA-L001", "SAGA-L101", fmt.Sprintf("unsupported escape \\%c", e), file, line, col}
					}
				} else {
					b.WriteRune(r)
				}
			}
			if l.end() {
				return nil, &SagaError{"SAGA-L001", "SAGA-L102", "unterminated string", file, line, col}
			}
			l.advance()
			out = append(out, Token{STRING, b.String(), line, col, file})
			continue
		}
		l.advance()
		add := func(k Kind, s string) { out = append(out, Token{k, s, line, col, file}) }
		switch c {
		case '(':
			add(LPAREN, "(")
		case ')':
			add(RPAREN, ")")
		case '{':
			add(LBRACE, "{")
		case '}':
			add(RBRACE, "}")
		case '[':
			add(LBRACKET, "[")
		case ']':
			add(RBRACKET, "]")
		case ',':
			add(COMMA, ",")
		case ':':
			add(COLON, ":")
		case ';':
			add(SEMICOLON, ";")
		case '@':
			add(AT, "@")
		case '?':
			add(QUESTION, "?")
		case '|':
			if l.match('>') {
				add(PIPE, "|>")
			} else {
				return nil, &SagaError{"SAGA-L001", "SAGA-L101", "unsupported character '|' (did you mean |>)", file, line, col}
			}
		case '+':
			add(PLUS, "+")
		case '%':
			add(PERCENT, "%")
		case '.':
			if l.match('.') {
				add(RANGE, "..")
			} else {
				add(DOT, ".")
			}
		case '-':
			if l.match('>') {
				add(ARROW, "->")
			} else {
				add(MINUS, "-")
			}
		case '*':
			if l.match('*') {
				add(POWER, "**")
			} else {
				add(STAR, "*")
			}
		case '/':
			add(SLASH, "/")
		case '=':
			if l.match('=') {
				add(EQEQ, "==")
			} else {
				add(EQUAL, "=")
			}
		case '!':
			if l.match('=') {
				add(BANGEQ, "!=")
			} else {
				add(BANG, "!")
			}
		case '<':
			if l.match('=') {
				add(LESSEQ, "<=")
			} else {
				add(LESS, "<")
			}
		case '>':
			if l.match('=') {
				add(GREATEREQ, ">=")
			} else {
				add(GREATER, ">")
			}
		default:
			return nil, &SagaError{"SAGA-L001", "SAGA-L101", fmt.Sprintf("unsupported character %q", c), file, line, col}
		}
	}
	out = append(out, Token{EOF, "", l.line, l.col, file})
	return out, nil
}
func (l *Lexer) end() bool { return l.pos >= len(l.src) }
func (l *Lexer) peek() rune {
	if l.end() {
		return 0
	}
	return l.src[l.pos]
}
func (l *Lexer) peekN(n int) rune {
	if l.pos+n >= len(l.src) {
		return 0
	}
	return l.src[l.pos+n]
}
func (l *Lexer) advance() rune {
	r := l.src[l.pos]
	l.pos++
	if r == '\n' {
		l.line++
		l.col = 1
	} else {
		l.col++
	}
	return r
}
func (l *Lexer) match(r rune) bool {
	if l.peek() != r {
		return false
	}
	l.advance()
	return true
}

func inRuneRanges(r rune, ranges []runeRange) bool {
	lo, hi := 0, len(ranges)
	for lo < hi {
		m := lo + (hi-lo)/2
		v := ranges[m]
		if r < v.lo {
			hi = m
		} else if r > v.hi {
			lo = m + 1
		} else {
			return true
		}
	}
	return false
}
func isStart(r rune) bool    { return r == '_' || inRuneRanges(r, xidStartRanges) }
func isContinue(r rune) bool { return r == '_' || inRuneRanges(r, xidContinueRanges) }
func isBidiControl(r rune) bool {
	switch r {
	case 0x061C, 0x200E, 0x200F, 0x202A, 0x202B, 0x202C, 0x202D, 0x202E, 0x2066, 0x2067, 0x2068, 0x2069:
		return true
	}
	return false
}

func decomposeRune(r rune, out *[]rune) {
	const (
		sBase  = 0xAC00
		lBase  = 0x1100
		vBase  = 0x1161
		tBase  = 0x11A7
		lCount = 19
		vCount = 21
		tCount = 28
		nCount = vCount * tCount
		sCount = lCount * nCount
	)
	if r >= sBase && r < sBase+sCount {
		sIndex := int(r - sBase)
		l := rune(lBase + sIndex/nCount)
		v := rune(vBase + (sIndex%nCount)/tCount)
		t := rune(tBase + sIndex%tCount)
		*out = append(*out, l, v)
		if t != tBase {
			*out = append(*out, t)
		}
		return
	}
	if seq, ok := canonicalDecomp[r]; ok {
		for _, p := range seq {
			decomposeRune(p, out)
		}
		return
	}
	*out = append(*out, r)
}
func composePair(a, b rune) (rune, bool) {
	const (
		sBase  = 0xAC00
		lBase  = 0x1100
		vBase  = 0x1161
		tBase  = 0x11A7
		lCount = 19
		vCount = 21
		tCount = 28
		nCount = vCount * tCount
		sCount = lCount * nCount
	)
	if a >= lBase && a < lBase+lCount && b >= vBase && b < vBase+vCount {
		return sBase + (a-lBase)*nCount + (b-vBase)*tCount, true
	}
	if a >= sBase && a < sBase+sCount && (a-sBase)%tCount == 0 && b > tBase && b < tBase+tCount {
		return a + (b - tBase), true
	}
	r, ok := canonicalCompose[[2]rune{a, b}]
	return r, ok
}
func normalizeNFC(text string) string {
	decomp := make([]rune, 0, len([]rune(text)))
	for _, r := range text {
		decomposeRune(r, &decomp)
	}
	for i := 1; i < len(decomp); i++ {
		c := combiningClass[decomp[i]]
		if c == 0 {
			continue
		}
		j := i
		for j > 0 {
			p := combiningClass[decomp[j-1]]
			if p == 0 || p <= c {
				break
			}
			decomp[j], decomp[j-1] = decomp[j-1], decomp[j]
			j--
		}
	}
	if len(decomp) == 0 {
		return ""
	}
	out := []rune{decomp[0]}
	starterPos := 0
	starter := decomp[0]
	lastCCC := uint8(0)
	for _, r := range decomp[1:] {
		ccc := combiningClass[r]
		if composed, ok := composePair(starter, r); ok && (lastCCC < ccc || lastCCC == 0) {
			out[starterPos] = composed
			starter = composed
			continue
		}
		out = append(out, r)
		if ccc == 0 {
			starterPos = len(out) - 1
			starter = r
		}
		lastCCC = ccc
	}
	return string(out)
}

File: saga/lexer.py
from __future__ import annotations

from decimal import Decimal

from .errors import LexError
from .tokens import KEYWORDS, Token, TokenKind
from .unicode_profile import (
    ensure_unicode_version, is_bidi_control, is_identifier_continue,
    is_identifier_start, validate_identifier,
)


class Lexer:
    def __init__(self, source: str, filename: str = "<input>") -> None:
        ensure_unicode_version()
        try:
            source_size = len(source.encode("utf-8", errors="strict"))
        except UnicodeEncodeError as exc:
            raise LexError("ソースコードに不正なUnicode文字があります", 1, 1, filename, detail_code="SAGA-L101") from exc
        # Saga source text has implementation-independent line semantics.
        # CRLF and legacy CR are normalized to LF before token positions are
        # calculated. A single leading UTF-8 BOM is permitted.
        source = source.replace("\r\n", "\n").replace("\r", "\n")
        if source.startswith("\ufeff"):
            source = source[1:]
        self.source = source
        self.filename = filename
        self.tokens: list[Token] = []
        self.start = 0
        self.current = 0
        self.line = 1
        self.column = 1
        self.start_line = 1
        self.start_column = 1

    def scan_tokens(self) -> list[Token]:
        while not self._at_end():
            self.start = self.current
            self.start_line = self.line
            self.start_column = self.column
            self._scan_token()
        self.tokens.append(Token(TokenKind.EOF, "", None, self.line, self.column, self.filename))
        return self.tokens

    def _scan_token(self) -> None:
        c = self._advance()
        single = {
            "(": TokenKind.LPAREN, ")": TokenKind.RPAREN,
            "{": TokenKind.LBRACE, "}": TokenKind.RBRACE,
            "[": TokenKind.LBRACKET, "]": TokenKind.RBRACKET,
            ",": TokenKind.COMMA, ":": TokenKind.COLON,
            ";": TokenKind.SEMICOLON, "+": TokenKind.PLUS,
            "%": TokenKind.PERCENT, "@": TokenKind.AT, "?": TokenKind.QUESTION,
        }
        if c in single:
            self._add(single[c]); return
        if c == ".":
            if self._match("."):
                self._add(TokenKind.RANGE); return
            self._add(TokenKind.DOT); return
        if c == "-":
            self._add(TokenKind.ARROW if self._match(">") else TokenKind.MINUS); return
        if c == "*":
            self._add(TokenKind.POWER if self._match("*") else TokenKind.STAR); return
        if c == "!":
            self._add(TokenKind.BANG_EQUAL if self._match("=") else TokenKind.BANG); return
        if c == "=":
            self._add(TokenKind.EQUAL_EQUAL if self._match("=") else TokenKind.EQUAL); return
        if c == "<":
            self._add(TokenKind.LESS_EQUAL if self._match("=") else TokenKind.LESS); return
        if c == ">":
            self._add(TokenKind.GREATER_EQUAL if self._match("=") else TokenKind.GREATER); return
        if c == "&":
            if self._match("&"):
                self._add(TokenKind.AND); return
            self._error("'&' は使えません", "論理積は and と書きます")
        if c == "|":
            if self._match(">"):
                self._add(TokenKind.PIPE); return
            if self._match("|"):
                self._add(TokenKind.OR); return
            self._error("'|' 単体は使えません", "パイプは |>、論理和は or と書きます")
        if c == "/":
            if self._match("/"):
                while self._peek() not in {"\n", "\0"}: self._advance()
            else:
                self._add(TokenKind.SLASH)
            return
        if c == "#":
            while self._peek() not in {"\n", "\0"}: self._advance()
            return
        if c in {" ", "\r", "\t", "\n"}: return
        if c in {'"', "'"}:
            self._string(c); return
        if "0" <= c <= "9":
            self._number(); return
        if c.isdigit():
            self._error(
                "数値リテラルの数字にはASCII 0-9を使用してください",
                "例: 123 または 3.14。Unicode数字は識別子との見分けを曖昧にするため数値構文では使用しません",
                "SAGA-L103",
            )
        if is_identifier_start(c):
            self._identifier(); return
        if is_bidi_control(c):
            self._error("双方向テキスト制御文字はソースコード中で使用できません", "文字列として必要な場合は\\u形式ではなく明示的なデータ入力を使用してください", "SAGA-L106")
        self._error(f"使用できない文字です: {c!r}")

    def _identifier(self) -> None:
        while is_identifier_continue(self._peek()):
            self._advance()
        text = self.source[self.start:self.current]
        error = validate_identifier(text)
        if error is not None:
            self._error(error, "同じ見た目の異なる識別子を避けるため、NFCの文字列を使用してください", "SAGA-L105" if "NFC" in error else "SAGA-L101")
        self._add(KEYWORDS.get(text, TokenKind.IDENT))

    def _number(self) -> None:
        while ("0" <= self._peek() <= "9") or self._peek() == "_": self._advance()
        kind = TokenKind.INT
        if self._peek() == "." and self._peek_next() == "_":
            self._error("小数点の直後に '_' は置けません", "例: 1.25 または 1.2_5", "SAGA-L103")
        if self._peek() == "." and self._peek_next() != "." and ("0" <= self._peek_next() <= "9"):
            kind = TokenKind.DECIMAL
            self._advance()
            while ("0" <= self._peek() <= "9") or self._peek() == "_": self._advance()
        raw = self.source[self.start:self.current]
        parts = raw.split(".")
        if any(not part or part.startswith("_") or part.endswith("_") or "__" in part for part in parts):
            self._error("数値の区切り '_' は数字の間に1つだけ置けます", "例: 1_000 または 3.141_592", "SAGA-L103")
        text = raw.replace("_", "")
        try:
            literal = Decimal(text) if kind is TokenKind.DECIMAL else int(text)
        except (ValueError, ArithmeticError):
            self._error("数値の書き方が正しくありません", diagnostic_id="SAGA-L103")
        self._add(kind, literal)

    def _string(self, quote: str) -> None:
        value: list[str] = []
        while not self._at_end():
            c = self._advance()
            if c == quote:
                self._add(TokenKind.STRING, "".join(value)); return
            if c == "\n": self._error("文字列は改行できません", "改行文字は \\n と書きます")
            if c == "\\":
                if self._at_end(): self._error("文字列のエスケープが途中で終わっています", diagnostic_id="SAGA-L102")
                escaped = self._advance()
                mapping = {"n": "\n", "t": "\t", "r": "\r", '"': '"', "'": "'", "\\": "\\"}
                if escaped not in mapping: self._error(f"未対応のエスケープです: \\{escaped}")
                value.append(mapping[escaped])
            else:
                value.append(c)
        self._error("文字列を閉じる引用符がありません", diagnostic_id="SAGA-L102")

    def _add(self, kind: TokenKind, literal: object | None = None) -> None:
        self.tokens.append(Token(
            kind, self.source[self.start:self.current], literal,
            self.start_line, self.start_column, self.filename,
        ))

    def _advance(self) -> str:
        c = self.source[self.current]; self.current += 1
        if c == "\n": self.line += 1; self.column = 1
        else: self.column += 1
        return c

    def _match(self, expected: str) -> bool:
        if self._at_end() or self.source[self.current] != expected: return False
        self._advance(); return True

    def _peek(self) -> str: return "\0" if self._at_end() else self.source[self.current]
    def _peek_next(self) -> str: return "\0" if self.current + 1 >= len(self.source) else self.source[self.current + 1]
    def _at_end(self) -> bool: return self.current >= len(self.source)

    def _error(self, message: str, hint: str | None = None, diagnostic_id: str | None = None) -> None:
        raise LexError(
            message, self.start_line, self.start_column, self.filename, hint,
            end_column=self.start_column + max(self.current - self.start, 1),
            detail_code=diagnostic_id,
        )
